File: leetcode/python/1232_check_if_it_is_a_straight_line.py
"""
T: O(N)
S: O(1)

Measure initial slope dy/dx. Check that the slope is the same for all
points. The regular way to do so is to check that dy/dx == dy_i/dx_i but that
can lead to division by zero. To avoid that we use multiplication form:
dx*dy_i == dx_i*dy
"""

from typing import List


class Solution:

    def checkStraightLine(self, coordinates: List[List[int]]) -> bool:
        (x1, y1), (x2, y2) = coordinates[:2]
        dx = x2 - x1
        dy = y2 - y1
        return all(dx * (y - y2) == (x - x2) * dy for x, y in coordinates)

File: leetcode/java/1232CheckIfItIsAStraightLine.java
/**
 * T: O(N) S: O(1)
 *
 * <p>Two points always form a line. Three points form a line if slope of the line between first and
 * second is the same as the slope between second and third. Since this property is associative, we
 * can expand that to arbitrarily many points by moving one point at a time and checking the last
 * three points to see if it still holds.
 */
class Solution {
  public boolean checkStraightLine(int[][] coordinates) {
    if (coordinates.length == 2) {
      return true;
    }
    for (int i = 0; i < coordinates.length - 2; i++) {
      if (!arePointsOnTheSameLine(coordinates, i)) {
        return false;
      }
    }
    return true;
  }

  private boolean arePointsOnTheSameLine(int[][] coordinates, int i) {
    int[] point0 = coordinates[i];
    int[] point1 = coordinates[i + 1];
    int[] point2 = coordinates[i + 2];
    int x0 = point0[0];
    int y0 = point0[1];
    int x1 = point1[0];
    int y1 = point1[1];
    int x2 = point2[0];
    int y2 = point2[1];
    return (x1 - x0) * (y2 - y1) == (x2 - x1) * (y1 - y0);
  }
}
